### PR TITLE
[VARC] Special-case hidden axes when applying RESET_UNSPECIFIED_AXES to pen.addVarComponent()

### PR DESCRIPTION
**Pull request by justvanrossum**
Alternative to #3541.

As explained in https://github.com/fonttools/fonttools/pull/3541#pullrequestreview-2088359154, the rawLocation mechanism is needed to meaningfully extract variable components via addVarComponent().

But this gives a bit of a conundrum with the new meaning of RESET_UNSPECIFIED_AXES.

Old meaning: when RESET_UNSPECIFIED_AXES is set, reset unspecified axes to default
New meaning: when RESET_UNSPECIFIED_AXES is set, reset unspecified axes the current font location

- to extract variable components for a glyph, we instantiate the glyph at all locations for which the glyph has variations
- this means we also need to specify values for the "private" axes used for variable components
- but _these_ values should not "leak" into descendants
- for normal rendering this is no problem, as the private axes are not user-settable (they are hidden), so the "global font location" for these axes is at default anyway

Long story short, to implement RESET_UNSPECIFIED_AXES in a useful way for addVarComponent(), we need to special case hidden axes.

### CODE REVIEW

**Comment by behdad** [comment]
> we need to special case hidden axes.

I'm a bit uncomfortable with that. Can we change the protocol and send two locations down?

**Comment by justvanrossum** [comment]
> I'm a bit uncomfortable with that. Can we change the protocol and send two locations down?

Hm, then we burden the addVarComponent signature with this problem. Not sure which is better.

The way we currently compile with RESET_UNSPECIFIED_AXES already depends on those axes always being at the default: as soon as you set one to a non-default value, the font breaks, as can be seen in (my private build of) FontGoggles:

https://github.com/fonttools/fonttools/assets/4246121/02b00d83-b20d-49dd-bdea-3102bd097207

So those axes are already special, and to me special-casing them here doesn't make things worse.

**Comment by behdad** [comment]
> Hm, then we burden the addVarComponent signature with this problem. Not sure which is better.

What if we add a lower-level `addVARCComponent` or something that takes exactly what's in the format? The location as well as the flag?

Let me also think more. In my compiler I drop encoding axes if they have the same value as the parent glyph. Let me think more about your requirements and get back to you.

**Comment by justvanrossum** [comment]
I'm having a hard time to come up with a use case for pen.addVarComponent() that is other than mine... There is of course the option that I don't use the pen protocol for this at all, and just read VARC directly. Maybe that makes more sense.

There's also your earlier idea if having _two_ flags:
1. reset to font location
2. reset to default

The first is handy when we pass down (user) axes to deeply nested components. The second is what we really need for private axes.

**Comment by justvanrossum** [comment]
(Hm, additionally, pen.addVarComponent() doesn't support extracting conditions (which Fontra doesn't yet support), but that also suggests that the pen protocol is perhaps not the right tool to extract variable components.)
